src/lib.rs

```diff
@@ -0,0 +1,1607 @@
+        )?;
+
+        ensure_active_opening_is_accepting_applications!(
+            active_stage,
+            DeactivateApplicationError::OpeningNotAcceptingApplications
+        )?;
+
+        // Ensure unstaking periods are OK.
+        ensure_opt_unstaking_period_is_ok!(
+            application_stake_unstaking_period,
+            opening.application_staking_policy,
+            DeactivateApplicationError::UnstakingPeriodTooShort(StakePurpose::Application),
+            DeactivateApplicationError::RedundantUnstakingPeriodProvided(StakePurpose::Application)
+        )?;
+
+        ensure_opt_unstaking_period_is_ok!(
+            role_stake_unstaking_period,
+            opening.role_staking_policy,
+            DeactivateApplicationError::UnstakingPeriodTooShort(StakePurpose::Role),
+            DeactivateApplicationError::RedundantUnstakingPeriodProvided(StakePurpose::Role)
+        )?;
+
+        //
+        // == MUTATION SAFE ==
+        //
+
+        // Deactive application
+        let result = Self::try_to_initiate_application_deactivation(
+            &application,
+            application_id,
+            application_stake_unstaking_period,
+            role_stake_unstaking_period,
+            &hiring::ApplicationDeactivationCause::External,
+        );
+
+        // MUST not have been ignored, is runtime invariant, false means code is broken.
+        // But should we do panic in runtime? Is there safer way?
+        assert_ne!(result, ApplicationDeactivationInitationResult::Ignored);
+
+        // DONE
+        Ok(())
+    }
+
+    /*
+    pub fn remove_application() ->() {
+
+        // Applies when an application is in the stage Inactive. Results in removing instance from applicationsById,
+        // and from the corresponding opening. Returns nothing.
+    }
+    */
+
+    /// The stake, with the given id, was unstaked.
+    pub fn unstaked(stake_id: T::StakeId) {
+        // Ignore unstaked
+        if !<ApplicationIdByStakingId<T>>::exists(stake_id) {
+            return;
+        }
+
+        // Get application
+        let application_id = <ApplicationIdByStakingId<T>>::get(stake_id);
+
+        assert!(<ApplicationById<T>>::exists(application_id));
+
+        let application = <ApplicationById<T>>::get(application_id);
+
+        // Make sure that we are actually unstaking, ignore otherwise.
+        let (deactivation_initiated, cause) = if let ApplicationStage::Unstaking {
+            deactivation_initiated,
+            cause,
+        } = application.stage
+        {
+            (deactivation_initiated, cause)
+        } else {
+            return;
+        };
+
+        //
+        // == MUTATION SAFE ==
+        //
+
+        // Drop stake from stake to application map
+        <ApplicationIdByStakingId<T>>::remove(stake_id);
+
+        // New values for application stakes
+        let new_active_role_staking_id = match application.active_role_staking_id {
```

Comment:
`map` is not applicable, because we don't always want to do `Some(id) => Some(f(id))`, specicially when we have `id == stake_id` we want `Some(id) => None`.